src/backend/distributed/commands/local_multi_copy.c

```diff
@@ -0,0 +1,208 @@
+		{
+			/*
+			 * We're going to flush the buffer to disk by effectively doing a full COPY command.
+			 * Hence we also need to add footers to the current buffer.
+			 */
+			AppendCopyBinaryFooters(localCopyOutState);
+		}
+		bool isEndOfCopy = false;
+		DoLocalCopy(localCopyOutState->fe_msgbuf, copyDest->distributedRelationId,
+					shardId,
+					copyDest->copyStatement, isEndOfCopy);
+	}
+}
+
+
+/*
+ * FinishLocalCopyToShard finishes local copy for the given shard with the shard id.
+ */
+void
+FinishLocalCopyToShard(CitusCopyDestReceiver *copyDest, int64 shardId,
+					   CopyOutState localCopyOutState)
+{
+	bool isBinaryCopy = localCopyOutState->binary;
+	if (isBinaryCopy)
+	{
+		AppendCopyBinaryFooters(localCopyOutState);
+	}
+	bool isEndOfCopy = true;
+	DoLocalCopy(localCopyOutState->fe_msgbuf, copyDest->distributedRelationId, shardId,
+				copyDest->copyStatement, isEndOfCopy);
+}
+
+
+/*
+ * AddSlotToBuffer serializes the given slot and adds it to the buffer in localCopyOutState.
+ */
+static void
+AddSlotToBuffer(TupleTableSlot *slot, CitusCopyDestReceiver *copyDest, CopyOutState
+				localCopyOutState)
+{
+	Datum *columnValues = slot->tts_values;
+	bool *columnNulls = slot->tts_isnull;
+	FmgrInfo *columnOutputFunctions = copyDest->columnOutputFunctions;
+	CopyCoercionData *columnCoercionPaths = copyDest->columnCoercionPaths;
+
+	AppendCopyRowData(columnValues, columnNulls, copyDest->tupleDescriptor,
+					  localCopyOutState, columnOutputFunctions,
+					  columnCoercionPaths);
+}
+
+
+/*
+ * ShouldSendCopyNow returns true if the given buffer size exceeds the
+ * local copy buffer size threshold.
+ */
+static bool
+ShouldSendCopyNow(StringInfo buffer)
+{
+	return buffer->len > LOCAL_COPY_FLUSH_THRESHOLD;
+}
+
+
+/*
+ * DoLocalCopy finds the shard table from the distributed relation id, and copies the given
+ * buffer into the shard.
+ * CopyFrom calls ReadFromLocalBufferCallback to read bytes from the buffer as though
+ * it was reading from stdin. It then parses the tuples and writes them to the shardOid table.
+ */
+static void
+DoLocalCopy(StringInfo buffer, Oid relationId, int64 shardId, CopyStmt *copyStatement,
+			bool isEndOfCopy)
+{
+	/*
+	 * Set the buffer as a global variable to allow ReadFromLocalBufferCallback to read from it.
+	 * We cannot pass additional arguments to ReadFromLocalBufferCallback.
+	 */
+	LocalCopyBuffer = buffer;
+
+	Oid shardOid = GetShardLocalTableOid(relationId, shardId);
+	Relation shard = heap_open(shardOid, RowExclusiveLock);
+	ParseState *pState = make_parsestate(NULL);
+
+	/* p_rtable of pState is set so that we can check constraints. */
+	pState->p_rtable = CreateRangeTable(shard, ACL_INSERT);
+
+	CopyState cstate = BeginCopyFrom(pState, shard, NULL, false,
+									 ReadFromLocalBufferCallback,
+									 copyStatement->attlist, copyStatement->options);
+	CopyFrom(cstate);
+	EndCopyFrom(cstate);
+
+	heap_close(shard, NoLock);
+	free_parsestate(pState);
+	resetStringInfo(buffer);
```

Comment:
It might be nicer to move this line into `WriteTupleToLocalShard`. Then it's clear that we're resetting `fe_msgbuf`.